### ISSUE TITLE
flex v2.6.4 build error with GCC 7.2.0

### ISSUE THREAD

**Issue by stefanosoffia** [spack/spack#6812]
The build of the [`flex` package](https://github.com/westes/flex) fails on Linux x64 with GCC 7.2.0.
This is a [known flex bug](https://github.com/westes/flex/issues/219) that will be fixed in flex v2.6.5, still [to be released](https://github.com/westes/flex/releases). 

I read [Spack packaging guide](http://spack.readthedocs.io/en/latest/packaging_guide.html), still I can't figure out what is the most appropriate solution to this issue. I guess something like black-listing flex v2.6.4 for this specific gcc version but I do not know how to implement it.

**Comment by adamjstewart** [comment]
> I guess something like black-listing flex v2.6.4 for this specific gcc version but I do not know how to implement it.

See the [`conflicts` directive](http://spack.readthedocs.io/en/latest/spack.html#spack.directives.conflicts). The [`meraculous` package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/meraculous/package.py#L42) is a good example of this.

Alternatively, if `flex` has released a patch to fix this bug, you could apply that as well.

**Comment by stefanosoffia** [comment]
Thanks for you clarification, not sure though about the result:
1. The `conflicts` directive results in a _concretization spec error_, i.e. something like the following:
 ```
  --- a/var/spack/repos/builtin/packages/flex/package.py
+++ b/var/spack/repos/builtin/packages/flex/package.py
@@ -33,6 +33,9 @@ class Flex(AutotoolsPackage):
     url = "https://github.com/westes/flex/releases/download/v2.6.1/flex-2.6.1.tar.gz"
 
     version('2.6.4', '2882e3179748cc9f9c23ec593d6adc8d')
+    # Build issue for v2.6.4 when gcc 7.2.0 is used.
+    # See issue #219; https://github.com/westes/flex/issues/219
+    conflicts('%gcc@7.2.0:', when='@2.6.4')
     version('2.6.3', 'a5f65570cd9107ec8a8ec88f17b31bb1')
     # Avoid flex '2.6.2' (major bug)
     # See issue #2554; https://github.com/westes/flex/issues/113
```
 results in an error like:
 ```
 ==> Error: Conflicts in concretized spec "flex@2.6.4%gcc@7.2.0+lex arch=linux-ubuntu17.10-x86_64 /skwrr53"

  List of matching conflicts for spec:

    flex@2.6.4%gcc@7.2.0+lex arch=linux-ubuntu17.10-x86_64 
        ^bison@3.0.4%gcc@7.2.0 patches=b72914fe38e54a6fc25f29019e0a0786705c4f61ce20d414cc2010c8d62448c7 ar
 ```
 That is, Spack does not fall back to the next flex version available, but it just aborts with an error. Am I missing something here?
2. The [proposed flex fix](https://github.com/westes/flex/issues/219) actually affects the generation of the `stage1flex` binary, which is not normally generated when building from the flex release archives, but which is instead included precompiled. In order to regenerate the culprit `stage1flex`, flex must be already available, i.e. one possible solution is to first build flex v2.6.3, which compiles fine with gcc 7.2, then to build from it 2.6.4.

**Comment by scheibelp** [comment]
> That is, Spack does not fall back to the next flex version available, but it just aborts with an error. Am I missing something here?

At the moment Spack does not use information provided by the `conflicts` directive to adjust the concretization process. You will have to explicitly specify properties on the command line to avoid reported conflicts, e.g. `spack install flex@2.6.4` if `gcc@7.2.0` is the default compiler.

**Comment by stefanosoffia** [comment]
Please see PR #6819

**Comment by adamjstewart** [comment]
Now that #6819 has been merged, can we close this issue?

**Comment by stefanosoffia** [comment]
Sorry to go back on this issue. Sure I can manually `spack install flex@2.6.3` but if my goal is to build another package which depends of `flex`, e.g. `pango -> gobject-introspection -> flex`, still Spack concretization process resolves `flex` to `flex@2.6.4`, i.e. the version which does not compile. Therefore the only option I have is to locally `spack edit flex` and remove the culprit version. Are there other solutions?

**Comment by adamjstewart** [comment]
There are a couple of other options. You can explicitly request the version of `flex` you want on the command line:
```
$ spack spec pango ^flex@2.6.3
```
or you can make this version of flex the default by adding/modifying your `packages.yaml` like so:
```yaml
packages:
  flex:
    version: [2.6.3]
```
Currently, there isn't anything in place for Spack to see that there is a conflict and automatically choose a different version. I think we can all agree that this would be the desired behavior, but changing the concretization implementation is harder than it sounds. Of course, pull requests to change this behavior would always be welcome, but I don't know enough about the concretizer to even point you in the right direction.
